File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/KeyPointCenter/KeyPoint/MatrixF.swift
//
//  MatrixF.swift
//  KeyPointCenter
//
//  Created by ZhaoTianyu on 2024/9/5.
//

import Foundation

public typealias MatrixA = [[Float]]

struct MatrixF {
    static func diagonal(columnCount: Int, value: Float) -> MatrixA {
        var matrix = MatrixA(repeating: [Float](repeating: 0, count: columnCount), count: columnCount)
        for i in 0..<columnCount {
            matrix[i][i] = value
        }
        return matrix
    }
    
//    static func add(_ a: MatrixA, _ b: MatrixA) -> MatrixA {
//        let rowCount = a.count
//        let columnCount = a[0].count
//        var result = Array(repeating: Array(repeating: Float(0.0), count: columnCount), count: rowCount)
//        for i in 0..<rowCount {
//            result.append(VectorF.add(a[i], b[i]))
//        }
//        return result
//    }
    
    static func add(_ a: MatrixA, _ b: MatrixA) -> MatrixA {
        if a.count != b.count || a[0].count != b[0].count {
            fatalError("Matrix dimensions must match")
        }
        
        let length = a.count
        let num2 = a[0].count
        var numArray = Array(repeating: Array(repeating: Float(0.0), count: num2), count: length)
        
        for i in 0..<num2 {
            for j in 0..<length {
                numArray[j][i] = a[j][i] + b[j][i]
            }
        }
        
        return numArray
    }
    
    static func sub(_ a: MatrixA, _ b: MatrixA) -> MatrixA {
        let rowCount = a.count
        let columnCount = a[0].count
        var result = Array(repeating: Array(repeating: Float(0.0), count: columnCount), count: rowCount)
        for i in 0..<rowCount {
            result.append(VectorF.sub(a[i], b[i]))
        }
        return result
    }
    
//    static func mul(_ a: MatrixA, _ b: MatrixA) -> MatrixA {
//        let rowCount = a.count
//        let columnCount = b[0].count
//        let commonDimension = a[0].count
//        var result = MatrixA(repeating: VectorA(repeating: 0, count: columnCount), count: rowCount)
//        for i in 0..<rowCount {
//            for j in 0..<columnCount {
//                for k in 0..<commonDimension {
//                    result[i][j] += a[i][k] * b[k][j]
//                }
//            }
//        }
//        return result
//    }
    static func mul(_ a: MatrixA, _ b: MatrixA) -> MatrixA {
        var result = Array(repeating: Array(repeating: Float(0.0), count: b[0].count), count: a.count)
        mul(a, b, result: &result)
        return result
    }
    
    static func mul(_ a: MatrixA, _ columnVector: [Float]) -> [Float] {
        let length = a.count
        if a[0].count != columnVector.count {
            fatalError("columnVector Vector must have the same length as columns in the matrix.")
        }
        var numArray = Array(repeating: Float(0.0), count: length)
        for i in 0..<length {
            for j in 0..<columnVector.count {
                numArray[i] += a[i][j] * columnVector[j]
            }
        }
        return numArray
    }
    
    static func mul(_ a: MatrixA, _ b: MatrixA, result: inout MatrixA) {
        let length = result.count
        let num2 = result[0].count
        var numArray = Array(repeating: Float(0.0), count: a[0].count)
        for i in 0..<num2 {
            for j in 0..<numArray.count {
                numArray[j] = b[j][i]
            }
            for k in 0..<length {
                var num6: Float = 0.0
                for m in 0..<numArray.count {
                    num6 += a[k][m] * numArray[m]
                }
                result[k][i] = num6
            }
        }
    }

    
    static func transpose(_ matrix: MatrixA) -> MatrixA {
        let rowCount = matrix.count
        let columnCount = matrix[0].count
        var result = MatrixA(repeating: VectorA(repeating: 0, count: rowCount), count: columnCount)
        
        for i in 0..<rowCount {
            for j in 0..<columnCount {
                result[j][i] = matrix[i][j]
            }
        }
        return result
    }
    
    static func inverse(_ matrix: MatrixA, _ inPlace: Bool = false) -> MatrixA? {
        let length = matrix.count
        let num2 = matrix[0].count
        if length != num2 {
            fatalError("Matrix must be square")
        }
        
        if length == 3 {
            let num3 = matrix[0][0]
            let num4 = matrix[0][1]
            let num5 = matrix[0][2]
            let num6 = matrix[1][0]
            let num7 = matrix[1][1]
            let num8 = matrix[1][2]
            let num9 = matrix[2][0]
            let num10 = matrix[2][1]
            let num11 = matrix[2][2]
            
            let num12 = ((num3 * ((num7 * num11) - (num8 * num10))) - (num4 * ((num6 * num11) - (num8 * num9)))) + (num5 * ((num6 * num10) - (num7 * num9)))
            
            if num12 == 0.0 {
                fatalError("Matrix is singular")
            }
            
            let num13 = 1.0 / num12
            var result = inPlace ? matrix : [[Float]](repeating: [Float](repeating: 0.0, count: 3), count: 3)
            
            result[0][0] = num13 * ((num7 * num11) - (num8 * num10))
            result[0][1] = num13 * ((num5 * num10) - (num4 * num11))
            result[0][2] = num13 * ((num4 * num8) - (num5 * num7))
            result[1][0] = num13 * ((num8 * num9) - (num6 * num11))
            result[1][1] = num13 * ((num3 * num11) - (num5 * num9))
            result[1][2] = num13 * ((num5 * num6) - (num3 * num8))
            result[2][0] = num13 * ((num6 * num10) - (num7 * num9))
            result[2][1] = num13 * ((num4 * num9) - (num3 * num10))
            result[2][2] = num13 * ((num3 * num7) - (num4 * num6))
            
            return result
        }
        
        if length != 2 {
            fatalError("Matrix size not supported")
        }
        
        let num14 = matrix[0][0]
        let num15 = matrix[0][1]
        let num16 = matrix[1][0]
        let num17 = matrix[1][1]
        let num18 = (num14 * num17) - (num15 * num16)
        
        if num18 == 0.0 {
            fatalError("Matrix is singular")
        }
        
        let num19 = 1.0 / num18
        var result = inPlace ? matrix : [[Float]](repeating: [Float](repeating: 0.0, count: 2), count: 2)
        
        result[0][0] = num19 * num17
        result[0][1] = -num19 * num15
        result[1][0] = -num19 * num16
        result[1][1] = num19 * num14
        
        return result
    }
}

File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/UnityDEBUG/ForDebug/Assets/Scripts/CharacterDriven/MatrixF.cs
using System;
public static class MatrixF
{
    // Methods
    public static float[,] Add(this float[,] a, float[,] b)
    {
        if ((a.GetLength(0) != b.GetLength(0)) || (a.GetLength(1) != b.GetLength(1)))
        {
            throw new ArgumentException("Matrix dimensions must match", "b");
        }
        int length = a.GetLength(0);
        int num2 = a.GetLength(1);
        int num1 = a.Length;
        float[,] numArray = new float[length, num2];
        for (int i = 0; i < num2; i++)
        {
            for (int j = 0; j < length; j++)
            {
                numArray[i, j] = a[i, j] + b[i, j];
            }
        }
        return numArray;
    }

    public static float[] Add(this float[] a, float[] b)
    {
        if (a == null)
        {
            throw new ArgumentNullException("a");
        }
        if (b == null)
        {
            throw new ArgumentNullException("b");
        }
        if (a.Length != b.Length)
        {
            throw new ArgumentException("Vector lengths must match", "b");
        }
        float[] numArray = new float[a.Length];
        for (int i = 0; i < a.Length; i++)
        {
            numArray[i] = a[i] + b[i];
        }
        return numArray;
    }

    public static int ColumnCount<T>(this T[,] matrix)
    {
        return matrix.GetLength(1);
    }

    public static T[,] Diagonal<T>(int size, T value)
    {
        T[,] localArray = new T[size, size];
        for (int i = 0; i < size; i++)
        {
            localArray[i, i] = value;
        }
        return localArray;
    }

    public static float[,] Identity(int size)
    {
        float[,] numArray = new float[size, size];
        for (int i = 0; i < size; i++)
        {
            numArray[i, i] = 1f;
        }
        return numArray;
    }

    public static float[,] Inverse(this float[,] matrix)
    {
        return matrix.Inverse(false);
    }

    public static float[,] Inverse(this float[,] matrix, bool inPlace)
    {
        int length = matrix.GetLength(0);
        int num2 = matrix.GetLength(1);
        if (length != num2)
        {
            throw new ArgumentException("Matrix must be square", "matrix");
        }
        if (length == 3)
        {
            float num3 = matrix[0, 0];
            float num4 = matrix[0, 1];
            float num5 = matrix[0, 2];
            float num6 = matrix[1, 0];
            float num7 = matrix[1, 1];
            float num8 = matrix[1, 2];
            float num9 = matrix[2, 0];
            float num10 = matrix[2, 1];
            float num11 = matrix[2, 2];
            float num12 = ((num3 * ((num7 * num11) - (num8 * num10))) - (num4 * ((num6 * num11) - (num8 * num9)))) + (num5 * ((num6 * num10) - (num7 * num9)));
            if (num12 == 0f)
            {
                throw new Exception();
            }
            float num13 = 1f / num12;
            float[,] singleArray1 = inPlace ? matrix : ((float[,])new float[3, 3]);
            singleArray1[0, 0] = num13 * ((num7 * num11) - (num8 * num10));
            singleArray1[0, 1] = num13 * ((num5 * num10) - (num4 * num11));
            singleArray1[0, 2] = num13 * ((num4 * num8) - (num5 * num7));
            singleArray1[1, 0] = num13 * ((num8 * num9) - (num6 * num11));
            singleArray1[1, 1] = num13 * ((num3 * num11) - (num5 * num9));
            singleArray1[1, 2] = num13 * ((num5 * num6) - (num3 * num8));
            singleArray1[2, 0] = num13 * ((num6 * num10) - (num7 * num9));
            singleArray1[2, 1] = num13 * ((num4 * num9) - (num3 * num10));
            singleArray1[2, 2] = num13 * ((num3 * num7) - (num4 * num6));
            return singleArray1;
        }
        if (length != 2)
        {
            throw new ArgumentException("Matrix not Support size", "matrix");
        }
        float num14 = matrix[0, 0];
        float num15 = matrix[0, 1];
        float num16 = matrix[1, 0];
        float num17 = matrix[1, 1];
        float num18 = (num14 * num17) - (num15 * num16);
        if (num18 == 0f)
        {
            throw new Exception();
        }
        float num19 = 1f / num18;
        float[,] singleArray2 = inPlace ? matrix : ((float[,])new float[2, 2]);
        singleArray2[0, 0] = num19 * num17;
        singleArray2[0, 1] = -num19 * num15;
        singleArray2[1, 0] = -num19 * num16;
        singleArray2[1, 1] = num19 * num14;
        return singleArray2;
    }

    public static float[,] Multiply(this float[,] a, float[,] b)
    {
        float[,] result = new float[a.GetLength(0), b.GetLength(1)];
        a.Multiply(b, result);
        return result;
    }

    public static float[] Multiply(this float[,] matrix, float[] columnVector)
    {
        int length = matrix.GetLength(0);
        if (matrix.GetLength(1) != columnVector.Length)
        {
            throw new Exception("columnVector Vector must have the same length as columns in the matrix.");
        }
        float[] numArray = new float[length];
        for (int i = 0; i < length; i++)
        {
            for (int j = 0; j < columnVector.Length; j++)
            {
                numArray[i] += matrix[i, j] * columnVector[j];
            }
        }
        return numArray;
    }

    public static void Multiply(this float[,] a, float[,] b, float[,] result)
    {
        int length = result.GetLength(0);
        int num2 = result.GetLength(1);
        float[] numArray = new float[a.GetLength(1)];
        for (int i = 0; i < num2; i++)
        {
            for (int j = 0; j < numArray.Length; j++)
            {
                numArray[j] = b[j, i];
            }
            for (int k = 0; k < length; k++)
            {
                float num6 = 0f;
                for (int m = 0; m < numArray.Length; m++)
                {
                    num6 += a[k, m] * numArray[m];
                }
                result[k, i] = num6;
            }
        }
    }

    public static float[,] Subtract(this float[,] a, float[,] b, bool inPlace = false)
    {
        if (a == null)
        {
            throw new ArgumentNullException("a");
        }
        if (b == null)
        {
            throw new ArgumentNullException("b");
        }
        if ((a.GetLength(0) != b.GetLength(0)) || (a.GetLength(1) != b.GetLength(1)))
        {
            throw new ArgumentException("Matrix dimensions must match", "b");
        }
        int length = a.GetLength(0);
        int num2 = b.GetLength(1);
        int num1 = a.Length;
        float[,] numArray = inPlace ? a : ((float[,])new float[length, num2]);
        for (int i = 0; i < length; i++)
        {
            for (int j = 0; j < num2; j++)
            {
                numArray[i, j] = a[i, j] - b[i, j];
            }
        }
        return numArray;
    }

    public static float[] Subtract(this float[] a, float[] b, bool inPlace = false)
    {
        if (a.Length != b.Length)
        {
            throw new ArgumentException("Vector length must match", "b");
        }
        float[] numArray = inPlace ? a : new float[a.Length];
        for (int i = 0; i < a.Length; i++)
        {
            numArray[i] = a[i] - b[i];
        }
        return numArray;
    }

    public static T[,] Transpose<T>(this T[,] matrix)
    {
        return matrix.Transpose<T>(false);
    }

    public static T[,] Transpose<T>(this T[,] matrix, bool inPlace)
    {
        int length = matrix.GetLength(0);
        int num2 = matrix.GetLength(1);
        if (inPlace)
        {
            if (length != num2)
            {
                throw new ArgumentException("Only square matrices can be transposed in place.", "matrix");
            }
            for (int j = 0; j < length; j++)
            {
                for (int k = j; k < num2; k++)
                {
                    T local = matrix[k, j];
                    matrix[k, j] = matrix[j, k];
                    matrix[j, k] = local;
                }
            }
            return matrix;
        }
        T[,] localArray = new T[num2, length];
        for (int i = 0; i < length; i++)
        {
            for (int m = 0; m < num2; m++)
            {
                localArray[m, i] = matrix[i, m];
            }
        }
        return localArray;
    }
}
